### PR TITLE
Calculate components and SHAP using empirical probabilities by default

### PR DESCRIPTION
**Pull request by jyliuu**
This PR includes changes to the `glex` function. 
- `glex` is now able to take ranger objects as input
- `glex` has a new optional parameter `probFunction` which specifies the probability function to use for weighting/marginalization of the leaves. If left to NULL, it will default to using the empirical marginal probabilities. Otherwise it can either be set to "path-dependent", i.e. using old weighting or to a custom probability function of the signature `function(coords, lb, ub)`
- Added some miscellaneous code for simulations under simulations/

### CODE REVIEW

**Review by jemus42 — COMMENTED** [review]
Should `simulation/` really be included in the repo? 
This seems like something that should go into a separate project rather than the package?

**Comment by MHiabu** [comment]
I agree that `simulation/` should be rather taken out.

File: R/glex.R [hunk 1]
```diff
@@ -15,6 +15,8 @@
 #'  Defaults to using all possible interactions available in the model.\cr
 #'  For [`xgboost`][xgboost::xgb.train], this defaults to the `max_depth` parameter of the model fit.\cr
 #'  If not set in `xgboost`, the default value of `6` is assumed.
+#' @param features Vector of column names in x to calculate components for. If \code{NULL}, all features are used.
```
Comment (jemus42) [inline R/glex.R:18]: Maybe explicitly state that the default is `NULL` -> all features

File: R/glex.R [hunk 2]
```diff
@@ -15,6 +15,8 @@
...
+#' @param probFunction Either "path-dependent" to use old path-dependent weighting of leaves or a user specified probability function of the signature function(coords, lb, ub). If left null, the empirical marginal probabilities will be used
```
Comment (jemus42) [inline R/glex.R:19]: Similarly here --- I usually try to make "what's the default value" the first thing in the parameter docs

File: R/glex.R [hunk 3]
```diff
@@ -115,66 +111,246 @@ glex.xgb.Booster <- function(object, x, max_interaction = NULL, ...) {
...
+# Function to get all subsets of set
+subsets <- function(x) {
+  if (length(x) == 1) {
+    list(integer(0), x)
+  } else {
+    do.call(c, lapply(0:length(x), combn, x = x, simplify = FALSE))
+  }
+}
+
+tree_fun_old <- function(tree, trees, x, all_S) {
```
Comment (jemus42) [inline R/glex.R:208]: Maybe rename this to something more informative? Functions with `_old` suffixes tend to just make me assume they're obsolete or deprecated, which doesn't appear to be the case here.

Thanks for prying out the internal functions though 😬

File: R/glex.R [hunk 4]
```diff
@@ -115,66 +111,246 @@ glex.xgb.Booster <- function(object, x, max_interaction = NULL, ...) {
...
-
-    # Return m matrix
-    m_all
-  }
 
   # Run in parallel if a parallel backend is registered
   j <- NULL
   idx <- 0:max(trees$Tree)
+
+  if (is.character(probFunction) && probFunction == "path-dependent")
```
Comment (jemus42) [inline R/glex.R:349]: Is it necessary to if/else here depending on `probFunction`, which is passed to `tree_fun_emp()` anyway if it's not equal to `"path-dependent"`? 

I mean, would it be simpler to have just one `tree_fun` that takes a `probFunction` argument and then internally switches behavior?

File: tests/testthat/test-glex-xgboost.R
Comment (jemus42) [inline tests/testthat/test-glex-xgboost.R:1]: These test now produce verbose output, flooding my console and the test pane:

```r
[1] "tree 0 of 49"
[1] "tree 1 of 49"
[1] "tree 2 of 49"
[1] "tree 3 of 49"
[1] "tree 4 of 49"
[1] "tree 5 of 49"
[1] "tree 6 of 49"
[1] "tree 7 of 49"
[1] "tree 8 of 49"
[1] "tree 9 of 49"
[1] "tree 10 of 49"
[1] "tree 11 of 49"
[1] "tree 12 of 49"
...
```

If this is debug output and no longer required please remove it, or if necessary put it behind some global option (e.g. `options(glex.verbose = TRUE)` or something) such that it is not shown by default.

**Review by jemus42 — CHANGES_REQUESTED** [review]
Cool 👍🏻 

Just a few things here and there for cleaning up.

File: R/glex.R (hunk 4 above)
Comment (jyliuu) [reply R/glex.R:348]: Well, the path_dependent tree_fun doesn't take a probFunction as argument, so that's the reason for the split

**Comment by jyliuu** [comment]
I have refactored the commit graph such that all commits related to `simulation/` have been removed 😅, and I have also committed your suggestions @jemus42. 

- Best regards

File: R/theme_glex.R
Comment (jemus42) [inline R/theme_glex.R:1]: Nice addition, thanks!

File: R/glex.R (hunk 4 above)
Comment (jemus42) [reply R/glex.R:348]: But that's my point - I was wondering if it was easier to have one `tree_fun()` that takes `probFunction` as an argument and then can either act as the path-dependent version (e.g. if `probFunction = NULL`?) or as currently `tree_fun_emp()` does.

I'm not saying it should be, just asking whether you think that would be easier/neater in the long run ¯\\\_(ツ)_/¯

File: debug.r
Comment (jemus42) [inline debug.r:1]: Thanks for removing the `simulation` bit but I think the `debug` bit also shouldn't be here right? 😅

File: R/glex.R (hunk 4 above)
Comment (jemus42) [reply R/glex.R:348]: Come to think of it: Regardless of how many `tree_fun_*`s there are, would you mind adding a basic roxygen skeleton for them? I know they are internal functions and not meant to be user-facing, but personally I think it is always helpful to have a few things documented for internal functions of a given complexity, especially input types and return value.

In this case for example I can mostly guess what the inputs are: `tree`, `trees`, `x`, `all_S` --- I can tell `trees` is a `data.table` and `tree` likely an integer but that's about it 🙃

File: NEWS.md
Comment (jemus42) [inline NEWS.md:1]: Almost forget: Don't forget to add your changes to the NEWS 😬

File: DESCRIPTION
Comment (jemus42) [inline DESCRIPTION:1]: Oh and please add yourself as an author (`role = "aut"`) here!

File: R/glex.R (hunk 4 above)
Comment (MHiabu) [reply R/glex.R:348]: Apart from that: `probFunction = "empirical"` should do the same as `probFunction = NULL` .
And if probFunction not in \{NULL, "empirical"\} or a prob function, it should give an error.

File: debug.r
Comment (jyliuu) [reply debug.r:1]: Yes, good catch :)

File: R/glex.R (hunk 4 above)
Comment (jyliuu) [reply R/glex.R:348]: Okay, I think I understand. I will just extract the if-statement into a separate function called tree-fun, which returns a closure (the actual treeFun) based on the specified probFunction.

File: DESCRIPTION
Comment (jyliuu) [reply DESCRIPTION:1]: Not sure how to do that haha

File: NEWS.md
Comment (jyliuu) [reply NEWS.md:1]: Should I then bump the version or?

**Review by MHiabu — APPROVED** [review]

File: NEWS.md
Comment (MHiabu) [reply NEWS.md:1]: good with me.

Comment (jemus42) [reply NEWS.md:1]: Yeah this is a minor version bump, 0.4.0

File: DESCRIPTION
Comment (jemus42) [reply DESCRIPTION:1]: See the `DESCRIPTION` file and add a line in the `Authors@R` field, similar to e.g. Munir's `person("Munir", "Hiabu", role = "aut"),` with your info, optionally with your ORCID if you want (see my entry in that field).

**Review by jemus42 — APPROVED** [review]

**Comment by jemus42** [comment]
Yeah turns out I messed up the versioning and told you the wrong thing, sorry.
This should bump glex 0.3.0.9000 -> glex 0.4.0 and I should tag a release, but don't worry about it I'll just do this minor maintenance stuff after merging this.

Anyway, sorry for the holdup, I'll merge this now, thanks! 👍🏻

**Pull request merged**